compiler/dyno/lib/uast/chpl-syntax-printer.cpp

```diff
@@ -0,0 +1,1065 @@
+    }
+
+    void visit(const Foreach* node) {
+      ss_ << "foreach ";
+
+      if (node->index()) {
+        printAst(ss_, node->index());
+        ss_ << " in ";
+      }
+      printAst(ss_, node->iterand());
+      ss_ << " ";
+      printBlockWithStyle(ss_, node->blockStyle(), node->stmts(), "do ");
+    }
+
+    void visit(const Formal* node) {
+      if (node->intent() != Formal::DEFAULT_INTENT) {
+        ss_ << kindToString(node->intent()) << " ";
+      }
+      ss_ << node->name().str();
+      if (const Expression* te = node->typeExpression()) {
+        ss_ << ": ";
+        printAst(ss_, te);
+      }
+      if (const Expression* ie = node->initExpression()) {
+        ss_ << " = ";
+        printAst(ss_, ie);
+      }
+    }
+
+    void visit(const ForwardingDecl* node) {
+      ss_ << "forwarding ";
+      if (node->expr()) {
+        printAst(ss_, node->expr());
+      }
+    }
+
+    void visit(const Function* node) {
+
+      printLinkages(node);
+
+      if (node->visibility() != Function::Visibility::DEFAULT_VISIBILITY) {
+        ss_ << kindToString(node->visibility()) << " ";
+      }
+
+      if (node->isOverride()) {
+        ss_ << "override ";
+      } else if (node->isInline()) {
+        ss_ << "inline ";
+      }
+
+      // Function Name
+      ss_ << kindToString(node->kind()) << " " << node->name().str();
+
+      // Formals
+      int numThisFormal = node->thisFormal() ? 1 : 0;
+      int nFormals = node->numFormals() - numThisFormal;
+      if (nFormals == 0 && node->isParenless()) {
+        // pass
+      } else if (nFormals == 0) {
+        ss_ << "()";
+      } else {
+        auto it = node->formals();
+        interpose(ss_, it.begin() + numThisFormal, it.end(), ", ", "(", ")");
+      }
+
+      // Return type
+      if (const Expression* e = node->returnType()) {
+        ss_ << " : ";
+        printAst(ss_, e);
+      }
+
+      // Return Intent
+      if (node->returnIntent() != Function::ReturnIntent::DEFAULT_RETURN_INTENT) {
+        ss_ << " " << kindToString(node->returnIntent());
+      }
+
+      interpose(ss_, node->stmts(), "\n", " {", "}");
+
+    }
+
+    void visit(const Identifier* node) {
+      ss_ << node->name().str();
+    }
+
+    void visit(const ImagLiteral* node) { return visitLiteral(ss_, node); }
+
+    void visit(const Import* node) {
+      if (node->visibility() != Decl::Visibility::DEFAULT_VISIBILITY) {
+        ss_ << kindToString(node->visibility());
+      }
+      ss_ << "import ";
+      interpose(ss_, node->visibilityClauses(), ", ");
+    }
+
+    void visit(const Include* node) {
+      ss_ << "include ";
+      if (node->visibility() != Decl::DEFAULT_VISIBILITY) {
+        ss_ << kindToString(node->visibility()) << " ";
+      }
+      if (node->isPrototype()) {
+        ss_ << "prototype ";
+      }
+      ss_ << "module ";
+      ss_ << node->name().c_str();
+    }
+
+    void visit(const IntLiteral* node)  { return visitLiteral(ss_, node); }
```

Comment:
You could farm these out to the parent class instead of specializing.

Reply:
will leave this for future work as it depends either making changes to `Literal.h` or keeping the specializations for `NumericLiteral<ValueT, ParamT>`